matchers/parse/lexer.go

```diff
@@ -0,0 +1,341 @@
+	return Position{
+		OffsetStart: l.start,
+		OffsetEnd:   l.pos,
+		ColumnStart: l.column,
+		ColumnEnd:   l.cols,
+	}
+}
+
+func (l *Lexer) scanOperator() (Token, error) {
+	if err := l.expect("!="); err != nil {
+		return Token{}, err
+	}
+
+	// Rewind because we need to know if the rune was an '!' or an '='.
+	l.rewind()
+
+	// If the first rune is an '!' then it must be followed with either an
+	// '=' or '~' to not match a string or regex.
+	if l.accept("!") {
+		if err := l.expect("=~"); err != nil {
+			return Token{}, err
+		}
+		return l.emit(TokenOperator), nil
+	}
+
+	// If the first rune is an '=' then it can be followed with an optional
+	// '~' to match a regex.
+	l.accept("=")
+	l.accept("~")
+	return l.emit(TokenOperator), nil
+}
+
+func (l *Lexer) scanQuoted() (Token, error) {
+	if err := l.expect("\""); err != nil {
+		return Token{}, err
+	}
+	var isEscaped bool
+	for r := l.next(); r != eof; r = l.next() {
+		if isEscaped {
+			isEscaped = false
+		} else if r == '\\' {
+			isEscaped = true
+		} else if r == '"' {
+			l.rewind()
+			break
+		}
+	}
+	if err := l.expect("\""); err != nil {
+		return Token{}, UnterminatedError{
+			input:       l.input,
+			offsetStart: l.start,
+			offsetEnd:   l.pos,
+			columnStart: l.column,
+			columnEnd:   l.cols,
+			quote:       '"',
+		}
+	}
+	return l.emit(TokenQuoted), nil
+}
+
+func (l *Lexer) scanUnquoted() (Token, error) {
+	for r := l.next(); r != eof; r = l.next() {
+		if isReserved(r) {
+			l.rewind()
+			break
+		}
+	}
+	return l.emit(TokenUnquoted), nil
+}
+
+func (l *Lexer) accept(valid string) bool {
+	if strings.ContainsRune(valid, l.next()) {
+		return true
+	}
+	l.rewind()
+	return false
+}
+
+// nolint:unused
+func (l *Lexer) acceptRun(valid string) {
+	for strings.ContainsRune(valid, l.next()) {
+	}
+	l.rewind()
+}
+
+func (l *Lexer) expect(valid string) error {
+	r := l.next()
+	if r == eof {
+		l.rewind()
+		return ExpectedError{
+			input:       l.input,
+			offsetStart: l.start,
+			offsetEnd:   l.pos,
+			columnStart: l.column,
+			columnEnd:   l.cols,
+			expected:    valid,
+		}
+	} else if !strings.ContainsRune(valid, r) {
+		l.rewind()
+		return ExpectedError{
+			input:       l.input,
+			offsetStart: l.start,
+			offsetEnd:   l.pos,
+			columnStart: l.column,
+			columnEnd:   l.cols,
+			expected:    valid,
+		}
+	} else {
+		return nil
+	}
+}
+
+func (l *Lexer) emit(kind TokenKind) Token {
+	tok := Token{
+		Kind:  kind,
+		Value: l.input[l.start:l.pos],
+		Position: Position{
+			OffsetStart: l.start,
+			OffsetEnd:   l.pos,
+			ColumnStart: l.column,
+			ColumnEnd:   l.cols,
+		},
+	}
+	l.start = l.pos
+	l.column = l.cols
+	return tok
+}
+
+func (l *Lexer) next() rune {
+	if l.pos >= len(l.input) {
+		l.width = 0
+		return eof
+	}
+	r, width := utf8.DecodeRuneInString(l.input[l.pos:])
+	l.width = width
+	l.pos += width
+	l.cols++
+	return r
+}
+
+func (l *Lexer) rewind() {
+	if l.width > 0 {
+		l.pos -= l.width
+		l.width = 0
+		l.cols--
+	}
```

Comment:
As discussed, I'm not sure I follow why we need `l.width > 0` can we document that?